[WIP]mgr/dashboard: monitoring: replace Grafana JSON with Grafonnet based code

aaSharma14 opened the pull request:
This PR intends to add grafonnet to generate grafana JSON files

Fixes: https://tracker.ceph.com/issues/45184
Signed-off-by: Aashish Sharma <aasharma@redhat.com>



<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

p-se commented on monitoring/grafana/dashboards/radosgw-sync-overview.jsonnet:
```diff
@@ -0,0 +1,66 @@
+  .addTargets([addTargetSchema('sum by (source_zone) (rate(ceph_data_sync_from_zone_fetch_bytes_sum[30s]))', 1, 'time_series', '{{source_zone}}')])
+             + {gridPos: {h: 7, w: 8, x: 0, y: 0}},
+
+  graphPanelSchema('Replication (objects) from Source Zone', 'null as zero', 'short', 'short', 'Objects/s', null, 1, '$Datasource')
```
Does it matter if `$Datasource` is used with a capitalized `D`? On L51 it is `$datasource`.

aaSharma14 replied:
Yeah, that's a typo. Will update that. Thanks @p-se

p-se commented:
It already looks like we'll be saving a lot of boilerplate code with this approach. Some of it is due to the data being written differently. In jsonnet there are some attributes on a single line, in JSON they aren't. But, assuming that `radosgw-sync-overview.jsonnet` in this PR is complete, it reduces the LOC from 440 in `radosgw-sync-voerview.json` to 66 in jsonnet. Which I think is significant. And we might even be able to reduce that a little bit more by outsourcing `grafana.annotation.datasource`.

p-se commented on monitoring/grafana/dashboards/radosgw-sync-overview.json:
```diff
@@ -1,440 +1,458 @@
-      },
-      "yaxes": [
-        {
-          "unit": "s",
```
Seems to have been removed as well. It might be okay, but we'll need to check.

aaSharma14 commented on monitoring/grafana/dashboards/test-jsonnet.sh:
```diff
@@ -0,0 +1,28 @@
+for file in $BASEDIR/jsonnet/*; do name=$(basename $file);
+jsonnet $file > ${TEMPDIR}/${name%.jsonnet}.json; done
+
+truncate -s 0 ${TEMPDIR}/json_differnce.log
```
Done. Thanks

aaSharma14 commented on monitoring/grafana/dashboards/test-jsonnet.sh:
```diff
@@ -0,0 +1,28 @@
+TEMPDIR=`mktemp -d`
+BASEDIR=$(dirname "$0")
+for file in $BASEDIR/jsonnet/*; do name=$(basename $file);
```
Done. Thanks

aaSharma14 commented:
> I've added a bunch of comments of things that I noticed and might have done differently. Those are not necessarily required to be changed, it's just my opinion.
> 
> But there a few things that we need to have a look at:
> 
> * the `jsondiff` tools seems to be a different one on my system (Ubuntu 20.04) and is not compatible with these changes
> * I'm not sure if it is safe to remove "iteration", but it might be. I hope it is, as it doesn't seem to be possible to keep it, is that correct?
> * a refresh time interval has likely been forgotten to be included
> * the `unit` field and its value has been removed in two panels. It might be okay, but we'll need to check.
> 
> Please consider the rest not as requested changes but as benevolent suggestions.

Thanks @p-se . I have addressed most of your comments. But as of now I don't see any side-effects of removing `iteration` and `unit` properties and also I couldn't yet find a way to add them as well. I have attached the screenshots for the RGW SYNC OVERVIEW grafana panel, before and after this PR. I can do more research on the iteration and unit properties and will update the PR accordingly.

p-se commented:
> > I've added a bunch of comments of things that I noticed and might have done differently. Those are not necessarily required to be changed, it's just my opinion.
> > But there a few things that we need to have a look at:
> > 
> > * the `jsondiff` tools seems to be a different one on my system (Ubuntu 20.04) and is not compatible with these changes
> > * I'm not sure if it is safe to remove "iteration", but it might be. I hope it is, as it doesn't seem to be possible to keep it, is that correct?
> > * a refresh time interval has likely been forgotten to be included
> > * the `unit` field and its value has been removed in two panels. It might be okay, but we'll need to check.
> > 
> > Please consider the rest not as requested changes but as benevolent suggestions.
> 
> Thanks @p-se . I have addressed most of your comments. But as of now I don't see any side-effects of removing `iteration` and `unit` properties and also I couldn't yet find a way to add them as well. I have attached the screenshots for the RGW SYNC OVERVIEW grafana panel, before and after this PR. I can do more research on the iteration and unit properties and will update the PR accordingly.

Yeah, might be the case that we don't need to worry about those.

Have you found anything out about the `jsondiff` issue? Is that  due to different versions of `jsondiff` or do I happen to have a different application installed? I'm asking because I'm not even how to resolve that issue for me, but I also wonder if we need to add another dependency to `install-dep.sh` or not.

aaSharma14 commented:
> > > I've added a bunch of comments of things that I noticed and might have done differently. Those are not necessarily required to be changed, it's just my opinion.
> > > But there a few things that we need to have a look at:
> > > 
> > > * the `jsondiff` tools seems to be a different one on my system (Ubuntu 20.04) and is not compatible with these changes
> > > * I'm not sure if it is safe to remove "iteration", but it might be. I hope it is, as it doesn't seem to be possible to keep it, is that correct?
> > > * a refresh time interval has likely been forgotten to be included
> > > * the `unit` field and its value has been removed in two panels. It might be okay, but we'll need to check.
> > > 
> > > Please consider the rest not as requested changes but as benevolent suggestions.
> > 
> > 
> > Thanks @p-se . I have addressed most of your comments. But as of now I don't see any side-effects of removing `iteration` and `unit` properties and also I couldn't yet find a way to add them as well. I have attached the screenshots for the RGW SYNC OVERVIEW grafana panel, before and after this PR. I can do more research on the iteration and unit properties and will update the PR accordingly.
> 
> Yeah, might be the case that we don't need to worry about those.
> 
> Have you found anything out about the `jsondiff` issue? Is that due to different versions of `jsondiff` or do I happen to have a different application installed? I'm asking because I'm not even how to resolve that issue for me, but I also wonder if we need to add another dependency to `install-dep.sh` or not.

I have updated the jsondiff command and removed `-s and -i` flags which didn't seem to be compatible for some distros. Are you still facing the same issue with jsondiff? 

> > > I've added a bunch of comments of things that I noticed and might have done differently. Those are not necessarily required to be changed, it's just my opinion.
> > > But there a few things that we need to have a look at:
> > > 
> > > * the `jsondiff` tools seems to be a different one on my system (Ubuntu 20.04) and is not compatible with these changes
> > > * I'm not sure if it is safe to remove "iteration", but it might be. I hope it is, as it doesn't seem to be possible to keep it, is that correct?
> > > * a refresh time interval has likely been forgotten to be included
> > > * the `unit` field and its value has been removed in two panels. It might be okay, but we'll need to check.
> > > 
> > > Please consider the rest not as requested changes but as benevolent suggestions.
> > 
> > 
> > Thanks @p-se . I have addressed most of your comments. But as of now I don't see any side-effects of removing `iteration` and `unit` properties and also I couldn't yet find a way to add them as well. I have attached the screenshots for the RGW SYNC OVERVIEW grafana panel, before and after this PR. I can do more research on the iteration and unit properties and will update the PR accordingly.
> 
> Yeah, might be the case that we don't need to worry about those.
> 
> Have you found anything out about the `jsondiff` issue? Is that due to different versions of `jsondiff` or do I happen to have a different application installed? I'm asking because I'm not even how to resolve that issue for me, but I also wonder if we need to add another dependency to `install-dep.sh` or not.

@p-se I have added the `jsondiff` dependency in `requirements.txt` file. Please let me know if it works fine for you now.

aaSharma14 commented:
jenkins test make check

aaSharma14 commented:
jenkins test dashboard

avanthakkar reviewed (approved):
LGTM! Great work @aaSharma14

alfonsomthd commented on install-deps.sh:
```diff
@@ -306,9 +306,11 @@ else
     [ $WITH_ZBD ] && with_zbd=true || with_zbd=false
     [ $WITH_PMEM ] && with_pmem=true || with_pmem=false
     source /etc/os-release
+    pip install jsondiff
```
Provided `jsondiff` is added in `requirements.txt`, I'm not sure we really need this.

epuertat replied:
I see a `python3-jsondiff` package for Ubuntu and Fedora, but since not available in CentOS, let's install from requirements.

epuertat commented on install-deps.sh:
```diff
@@ -306,9 +306,11 @@ else
     case "$ID" in
     debian|ubuntu|devuan|elementary)
         echo "Using apt-get to install dependencies"
+        $SUDO apt-get install -y jsonnet
```
Also this needs to be added to the `debian/control` file instead.

aaSharma14 replied:
Done

epuertat commented on install-deps.sh:
```diff
@@ -364,6 +366,7 @@ else
         case "$ID" in
             fedora)
                 $SUDO dnf install -y dnf-utils
+                $SUDO dnf install -y golang-github-google-jsonnet
```
Let's add this dependency to the ceph.spec.in, in the BuildRequires instead. If it has different names for Fedora and CentOS you can use some macros the check whether it's Fedora or CentOS.

aaSharma14 replied:
Done

epuertat commented on monitoring/grafana/dashboards/generate-jsonnet.sh:
```diff
@@ -0,0 +1,9 @@
+set -e
+BASEDIR=$(dirname "$0")
+
+for file in $BASEDIR/jsonnet/*; do
+    name=$(basename $file)
+    jsonnet $file > $BASEDIR/${name%.jsonnet}.json
+done
```
It seems jsonnet supports generating multiple files from the single jsonnet file. Have you considered that? Is it worthy?

aaSharma14 replied:
Thanks @epuertat . I have updated the PR with generating multiple json files from single jsonnet file. PTAL.

aaSharma14 commented on .gitmodules:
```diff
@@ -78,3 +78,7 @@
 	path = src/libkmip
 	url = https://github.com/ceph/libkmip
 	branch = ceph-master
+[submodule "grafonnet-lib"]
+	path = grafonnet-lib
+	url = https://github.com/grafana/grafonnet-lib.git
+	branch = ceph-master
```
The other option is to install the [jsonnet-bundler](https://grafana.github.io/grafonnet-lib/getting-started/) package. But I couldn't find any package support for jsonnet-bundler in other distros except fedora - https://pkgs.org/search/?q=jsonnet-bundler

aaSharma14 replied:
Done and updated

p-se reviewed (approved):
lgtm, however, would you mind adding a note to the [Ceph Dashboard docs](https://docs.ceph.com/en/latest/mgr/dashboard/#enabling-the-embedding-of-grafana-dashboards) that the Grafana dashboards cannot be simply downloaded from Github anymore but must be generated (at least one needs to be generated at this point, but eventually I'd be all of them).

epuertat commented:
> lgtm, however, would you mind adding a note to the [Ceph Dashboard docs](https://docs.ceph.com/en/latest/mgr/dashboard/#enabling-the-embedding-of-grafana-dashboards) that the Grafana dashboards cannot be simply downloaded from Github anymore but must be generated (at least one needs to be generated at this point, but eventually I'd be all of them).

@p-se but the JSON files are still there. We're changing how they're generated not when. What we might need to document is some advice/hints on how to modify them (@aaSharma14 based on your experience): work directly in Grafana, get the resulting JSON and update the corresponding jsonnet?

epuertat commented on monitoring/grafana/dashboards/test-jsonnet.sh:
```diff
@@ -0,0 +1,31 @@
+    do
+        GENERATED_FILE_NAME=$(basename $generated_files)
+        if [ $JSON_FILE_NAME == $GENERATED_FILE_NAME ]; then
+            jsondiff --indent 2 $generated_files $json_files >> ${TEMPDIR}/json_difference.log
+            jsondiff --indent 2 $generated_files $json_files
+        fi
+    done
+done
+
+if grep -Fxq "{}" ${TEMPDIR}/json_difference.log
+then
+        rm -rf ${TEMPDIR}
+        echo "Congratulations! Grafonnet Check Passed"
+else
+        rm -rf ${TEMPDIR}
+        echo "Grafonnet Check Failed"
+        exit 1
+fi
```
You can heavily simplify all of this with:
```bash
error=0
if [ "$(jsondiff a.json c.json | tee --append ${TEMPDIR}/json_difference.log )" != "{}" ]
then
  echo "Grafonnet error in file: $..."
  error=1
fi

...

exit $error
```

epuertat commented on src/pybind/mgr/dashboard/tox.ini:
```diff
@@ -6,7 +6,8 @@ envlist =
     apidocs,
-    openapi-{check,fix, doc}
+    openapi-{check,fix, doc},
+    grafonnet-check
```
You should provide both the check and fix actions, same as with openapi. The fix is the generate one (it simply overwrites the existing json).

aaSharma14 replied:
Sure, will do that.

epuertat replied:
You need to expose the -fix here too, or tox won't allow you to call it.

aaSharma14 replied:
Sure

epuertat reviewed (changes requested):
Good progress!

p-se commented on monitoring/grafana/dashboards/test-jsonnet.sh:
```diff
@@ -0,0 +1,31 @@
+TEMPDIR=`mktemp -d`
+BASEDIR=$(dirname "$0")
+
+jsonnet -m ${TEMPDIR} $BASEDIR/jsonnet/grafana_dashboards.jsonnet
```
I've simplified the script and added a feature which will print a traceback and exit with 1 if any of the JSON files invalid. That output was suppressed due to `set -e` (as I suggested). You may use (and adapt) it, if you want.

```bash
#!/usr/bin/env bash

# Run this script in src/pybind/mgr/dashboard with `tox -e grafonnet` if you
# have any issues running this script directly.

set -e

tmpdir=$(mktemp -d)
clean() {
    rm ${tmpdir} -rf
}
trap clean 0
trap clean 1

cd $(dirname "$0")

for gen_json in $(jsonnet -m ${tmpdir} jsonnet/grafana_dashboards.jsonnet); do
    base_name=$(basename $gen_json)
    # jsondiff exits with exit code 1 if both, there's a difference between two
    # files AND if it throws a Traceback due to the JSON of one file isn't
    # valid, so we check the output instead of the exit code.
    diff=$(jsondiff --indent 2 $gen_json $base_name 2>&1 || true)

    if [[ "$diff" =~ ^Traceback ]]; then
        echo -e "$diff"
        exit 1
    elif [[ "$diff" != "{}" ]] ; then
        echo "failed comparing generated file with existing, please re-generate $base_name"
        echo "$diff"
        exit 1
    fi
done

echo "success"
```

aaSharma14 replied:
This script works better, Thanks for this @p-se .

aaSharma14 commented on monitoring/grafana/dashboards/generate-jsonnet.sh:
```diff
@@ -0,0 +1,6 @@
+#!/usr/bin/env bash
+
+set -e
+BASEDIR=$(dirname "$0")
+
+jsonnet -m $BASEDIR $BASEDIR/jsonnet/grafana_dashboards.jsonnet
```
This script is to explicitly generate the dashboard json files from jsonnet file if required. I will remove the script and add this to grafonnet-fix in tox.ini.

aaSharma14 commented on monitoring/grafana/dashboards/jsonnet/grafana_dashboards.jsonnet:
```diff
@@ -0,0 +1,73 @@
+    graphPanel.new(title=title,nullPointMode=nullPointMode,stack=stack,formatY1=formatY1,formatY2=formatY2,labelY1=labelY1,labelY2=labelY2,min=min,fill=fill,datasource=datasource);
+    local addTargetSchema(expr, intervalFactor, format, legendFormat) =
+    prometheus.target(expr=expr,intervalFactor=intervalFactor,format=format,legendFormat=legendFormat);
+    local addTemplateSchema(name, datasource, query, refresh, hide, includeAll, sort) =
```
No, each dashboard has different set of values for this.

aaSharma14 replied:
Not actually, these values vary across different dashboards.

aaSharma14 commented on monitoring/grafana/dashboards/jsonnet/grafana_dashboards.jsonnet:
```diff
@@ -0,0 +1,73 @@
+
+    graphPanelSchema('Replication (objects) from Source Zone', 'null as zero', true, 'short', 'short', 'Objects/s', null, 0, 1, '$datasource')
+    .addTargets([addTargetSchema('sum by (source_zone) (rate(ceph_data_sync_from_zone_fetch_bytes_count[30s]))', 1, 'time_series', '{{source_zone}}')])
+                + {gridPos: {h: 7, w: 7.4, x: 8.3, y: 0}},
```
Yes, these are the existing set of values used here. I will check if it can be 8 and 8.

aaSharma14 commented on monitoring/grafana/dashboards/jsonnet/grafana_dashboards.jsonnet:
```diff
@@ -0,0 +1,73 @@
+    )
+    )
+
+    .addPanels([
+    graphPanelSchema('Replication (throughput) from Source Zone', 'null as zero', true, 'Bps', 'short', null, null, 0, 1, '$datasource')
+    .addTargets([addTargetSchema('sum by (source_zone) (rate(ceph_data_sync_from_zone_fetch_bytes_sum[30s]))', 1, 'time_series', '{{source_zone}}')])
+                + {gridPos: {h: 7, w: 8, x: 0, y: 0}},
+
+    graphPanelSchema('Replication (objects) from Source Zone', 'null as zero', true, 'short', 'short', 'Objects/s', null, 0, 1, '$datasource')
+    .addTargets([addTargetSchema('sum by (source_zone) (rate(ceph_data_sync_from_zone_fetch_bytes_count[30s]))', 1, 'time_series', '{{source_zone}}')])
+                + {gridPos: {h: 7, w: 7.4, x: 8.3, y: 0}},
+
+    graphPanelSchema('Polling Request Latency from Source Zone', 'null as zero', true, 'ms', 'short', null, null, 0, 1, '$datasource')
+    .addTargets([addTargetSchema('sum by (source_zone) (rate(ceph_data_sync_from_zone_poll_latency_sum[30s]) * 1000)', 1, 'time_series', '{{source_zone}}')])
+                + {gridPos: {h: 7, w: 8, x: 16, y: 0}},
+
+    graphPanelSchema('Unsuccessful Object Replications from Source Zone', 'null as zero', true, 'short', 'short', 'Count/s', null, 0, 1, '$datasource')
+    .addTargets([addTargetSchema('sum by (source_zone) (rate(ceph_data_sync_from_zone_fetch_errors[30s]))', 1, 'time_series', '{{source_zone}}')])
+                + {gridPos: {h: 7, w: 8, x: 0, y: 7}},           
+    ])
```
Thanks @epuertat . WIll do that.

aaSharma14 commented on monitoring/grafana/dashboards/jsonnet/grafana_dashboards.jsonnet:
```diff
@@ -0,0 +1,73 @@
+    schemaVersion=16,
+    tags=["overview"],
+    timezone='',
+    timepicker={refresh_intervals:['5s','10s','15s','30s','1m','5m','15m','30m','1h','2h','1d'],time_options:['5m','15m','1h','6h','12h','24h','2d','7d','30d']}
```
Sure

p-se commented on monitoring/grafana/dashboards/jsonnet/grafana_dashboards.jsonnet:
```diff
@@ -0,0 +1,73 @@
+    timepicker={refresh_intervals:['5s','10s','15s','30s','1m','5m','15m','30m','1h','2h','1d'],time_options:['5m','15m','1h','6h','12h','24h','2d','7d','30d']}
+    )
+
+    .addAnnotation(
+    grafana.annotation.datasource(
+        builtIn= 1,
+        datasource= '-- Grafana --',
+        enable= true,
+        hide= true,
+        iconColor= 'rgba(0, 211, 255, 1)',
+        name= 'Annotations & Alerts',
+        type= 'dashboard'
+    )
+    )
```
Hm, I didn't know, but now I've looked it up. It seems we can do things like showing Prometheus alerts in Grafana graphs using annotations. We'd need to think about where to put them, because I don't think we want to have them in every panel. But at the right place, that might be really nice. However, that's probably not something to be added in this PR.

djgalloway commented on install-deps.sh:
```diff
@@ -306,6 +306,11 @@ else
+    # check if grafonnet-lib is there else download
+    if [ ! -d "grafonnet-lib-0.1.0" ];
+    then
+        $SUDO wget -c https://github.com/grafana/grafonnet-lib/archive/refs/tags/v0.1.0.tar.gz -O - | sudo tar -xz
```
Why `$SUDO`?  If it's absolutely necessary, please `chown` the directory afterwards.

This PR is causing the Jenkins builders to fail a bunch of other jobs because the Jenkins user can't delete that directory.

e.g., https://jenkins.ceph.com/job/ceph-api/22209/console

aaSharma14 replied:
Updated the PR with this change. Sorry for the inconvenience @djgalloway .

tchaikov commented on install-deps.sh:
```diff
@@ -306,6 +306,11 @@ else
     [ $WITH_ZBD ] && with_zbd=true || with_zbd=false
     [ $WITH_PMEM ] && with_pmem=true || with_pmem=false
     source /etc/os-release
+    # check if grafonnet-lib is there else download
+    if [ ! -d "grafonnet-lib-0.1.0" ];
+    then
+        wget -c https://github.com/grafana/grafonnet-lib/archive/refs/tags/v0.1.0.tar.gz -O - | sudo tar -xz
+    fi
```
why sudo ? i'd suggest avoid dumping code into `install-deps.sh` if unnecessary . instead i'd recommend `ExternalProject_Add()` cmake command. it took me quite a while to fix 53a5816deda0874a3a37e131e9bc22d88bb2a588 ..

tchaikov replied:
```
Wiping out workspace first.
Also:   java.nio.file.FileSystemException: /home/jenkins-build/build/workspace/ceph-api/grafonnet-lib-0.1.0/.circleci/config.yml: Operation not permitted
		at sun.nio.fs.UnixException.translateToIOException(UnixException.java:100)
		at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:111)
		at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:116)
```

i think this change breaks all the jenkins jobs which are scheduled by this PR.

djgalloway replied:
Addressed [here](https://github.com/ceph/ceph/pull/42194#discussion_r678416834).

I'll re-run my one-liner to delete that dir on all builders.

epuertat replied:
@tchaikov: @aaSharma14 initially installed grafonnet as a git submodule, but I recommended him to use an alternative approach (I'm not a fan of submodules), like downloading the tarball. What's your take on that?

epuertat replied:
BTW @tchaikov the downside of that approach you suggest is that you need internet connection during the checks phase, right? I thought that's why we wanted to prefetch all the deps...

tchaikov replied:
i think we already need internet connection during `make check`, for preparing the npm environment and for downloading nodejs packages.

tchaikov replied:
> BTW @tchaikov the downside of that approach you suggest is that you need internet connection during the checks phase, right? I thought that's why we wanted to prefetch all the deps...

might want to use "URL Download". see https://cmake.org/cmake/help/latest/module/ExternalProject.html

epuertat replied:
Looks good @tchaikov !

epuertat commented on monitoring/grafana/dashboards/jsonnet/grafana_dashboards.jsonnet:
```diff
@@ -0,0 +1,42 @@
+    .addPanels([
+    RgwSyncOverviewPanel('Replication (throughput) from Source Zone', 'Bps', null, 'ceph_data_sync_from_zone_fetch_bytes_sum', '', 7, 8, 0, 0),
+    RgwSyncOverviewPanel('Replication (objects) from Source Zone', 'short', 'Objects/s', 'ceph_data_sync_from_zone_fetch_bytes_count', '', 7, 7.4, 8.3, 0),
+    RgwSyncOverviewPanel('Polling Request Latency from Source Zone', 'ms', null, 'ceph_data_sync_from_zone_poll_latency_sum', ' * 1000', 7, 8, 16, 0),
```
Probably you can remove the RgwPower param since Grafana should autoscale the chart if we use seconds... Isn't that the case?

epuertat commented on src/pybind/mgr/dashboard/tox.ini:
```diff
@@ -176,6 +177,20 @@ commands =
+    check: Ensure that auto-generated grafana dashboard files matches the current version
+    fix: generate dashboard json files from jsonnet file with latest changes
+deps =
+    {[base]deps}
```
Do we really need all the python deps for this? Maybe it makes more sense to install the only python dep we need here from a separate `requirements-grafonnet.txt`?

aaSharma14 replied:
Sure

epuertat reviewed (changes requested):
Looks better! Thanks @aaSharma14 for addressing my comments.

aaSharma14 commented on install-deps.sh:
```diff
@@ -306,6 +306,11 @@ else
     [ $WITH_ZBD ] && with_zbd=true || with_zbd=false
     [ $WITH_PMEM ] && with_pmem=true || with_pmem=false
     source /etc/os-release
+    # check if grafonnet-lib is there else download
```
Yes, it is required to generate the grafana dashboard jsons from jsonnet file in make check.

aaSharma14 commented:
jenkins retest this please

tchaikov commented on src/pybind/mgr/dashboard/CMakeLists.txt:
```diff
@@ -4,6 +4,13 @@ install(DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}
   PATTERN ".*" EXCLUDE)
 
+include(ExternalProject)
+ExternalProject_Add(grafonnet-lib-0.1.0
```
please do not encode the version in the target name unless the version really matters like python2 and python3.

how is this target integrated into the build process?

epuertat replied:
@tchaikov I was initially not keen on pinning deps versions, but with [the increasing number of malware being injected into code or package repos](https://www.bleepingcomputer.com/news/security/pypi-packages-caught-stealing-credit-card-numbers-discord-tokens/), I think version pinning is the simplest way to avoid [supply chain attacks](https://www.wired.com/story/hacker-lexicon-what-is-a-supply-chain-attack/). This is the same we are doing with the Github Actions (pinning to specific commits we have manually reviewed).

tchaikov replied:
@epuertat could you please re-read my comment?

epuertat replied:
Ok, your concern was about adding the version IN the target name, not about the version pinning itself, right?

tchaikov replied:
right.

epuertat commented on monitoring/grafana/dashboards/jsonnet/grafana_dashboards.jsonnet:
```diff
@@ -0,0 +1,42 @@
+
+    .addRequired(type= 'grafana',id= 'grafana',name= 'Grafana',version= '5.0.0')
+
+    .addRequired(type= 'panel',id= 'graph',name= 'Graph',version= '5.0.0')
```
Are you sure the linter/formatter is ok with this. Usually whitespaces happen after commas, not after = or other binary operators. I see this style issue happening a lot, and maybe it's just me, but it makes it hard for me to read the jsonnet code. 

```suggestion
    .addRequired(type='panel', id='graph', name='Graph', version='5.0.0')
```

Can you please, if not too complex, add a [linter for jsonnet](https://jsonnet.org/learning/tools.html)?

epuertat commented on monitoring/grafana/dashboards/test-jsonnet.sh:
```diff
@@ -0,0 +1,31 @@
+    do
+        GENERATED_FILE_NAME=$(basename $generated_files)
+        if [ $JSON_FILE_NAME == $GENERATED_FILE_NAME ]; then
+            jsondiff --indent 2 $generated_files $json_files >> ${TEMPDIR}/json_difference.log
+            jsondiff --indent 2 $generated_files $json_files
```
When you want to save some output to a file while still printing it to the screen, `tee` is your friend:
```sh
jsondiff --indent 2 $generated_files $json_files | tee -a ${TEMPDIR}/json_difference.log
```

epuertat commented on src/pybind/mgr/dashboard/CMakeLists.txt:
```diff
@@ -4,12 +4,20 @@ install(DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}
   PATTERN ".*" EXCLUDE)
 
+include(ExternalProject)
+ExternalProject_Add(grafonnet-lib
+  DOWNLOAD_DIR ${CMAKE_BINARY_DIR}
+  GIT_REPOSITORY "https://github.com/grafana/grafonnet-lib.git"
+  GIT_TAG "v0.1.0"
+  GIT_SHALLOW TRUE)
```
👍🏼

epuertat commented on src/pybind/mgr/dashboard/tox.ini:
```diff
@@ -6,7 +6,8 @@ envlist =
     apidocs,
-    openapi-{check,fix, doc}
+    openapi-{check,fix, doc},
+    grafonnet-{check,fix}
```
Thanks a lot!

epuertat commented on monitoring/grafana/dashboards/test-jsonnet.sh:
```diff
@@ -0,0 +1,31 @@
+    do
+        GENERATED_FILE_NAME=$(basename $generated_files)
+        if [ $JSON_FILE_NAME == $GENERATED_FILE_NAME ]; then
+            jsondiff --indent 2 $generated_files $json_files >> ${TEMPDIR}/json_difference.log
```
Maybe we should add the linter here too?

epuertat reviewed (changes requested):
I'm almost sure this is the last set of changes I request! Good job!

aaSharma14 commented on monitoring/grafana/dashboards/jsonnet/grafana_dashboards.jsonnet:
```diff
@@ -0,0 +1,42 @@
+local grafana = import '../../../../grafonnet-lib/grafonnet/grafana.libsonnet';
```
@tchaikov I am getting an error in the make-check 'couldn't open import "../../../../grafonnet-lib/grafonnet/grafana.libsonnet": no match locally or in the Jsonnet library paths.'. Maybe am doing something wrong here as the jsonnet file is not being able to locate the grafonnet-lib directory. I used `include_directories(grafonnet-lib ${CMAKE_BINARY_DIR}})` and tried `add_dependencies` as well. Can you please suggest here some solution.

epuertat commented on monitoring/grafana/dashboards/jsonnet/grafana_dashboards.jsonnet:
```diff
@@ -0,0 +1,42 @@
+local dashboard = grafana.dashboard;
+local template = grafana.template;
+local graphPanel = grafana.graphPanel;
+local prometheus = grafana.prometheus;
+local annotation = grafana.annotation;
+local tablePanel = grafana.tablePanel;
+local dashboardSchema(title, uid, time_from, refresh, schemaVersion, tags,timezone, timepicker) =
+dashboard.new(title=title,uid=uid,time_from=time_from,refresh=refresh,schemaVersion=schemaVersion,tags=tags,timezone=timezone,timepicker=timepicker);
+local graphPanelSchema(title, nullPointMode, stack, formatY1, formatY2, labelY1, labelY2, min, fill, datasource) =
+graphPanel.new(title=title,nullPointMode=nullPointMode,stack=stack,formatY1=formatY1,formatY2=formatY2,labelY1=labelY1,labelY2=labelY2,min=min,fill=fill,datasource=datasource);
+local addTargetSchema(expr, intervalFactor, format, legendFormat) =
+prometheus.target(expr=expr,intervalFactor=intervalFactor,format=format,legendFormat=legendFormat);
+local addTemplateSchema(name, datasource, query, refresh, hide, includeAll, sort) =
+template.new(name=name,datasource=datasource,query=query,refresh=refresh,hide=hide,includeAll=includeAll, sort=sort);
+local addAnnotationSchema(builtIn, datasource, enable, hide, iconColor, name, type) =
+grafana.annotation.datasource(builtIn=builtIn,datasource=datasource,enable=enable,hide=hide,iconColor=iconColor,name=name,type=type);
+{
+  "radosgw-sync-overview.json":
+    local RgwSyncOverviewPanel(title, formatY1, labelY1, rgwMetric, x, y, w, h) =
+    graphPanelSchema(title, 'null as zero', true, formatY1, 'short', labelY1, null, 0, 1, '$datasource')
+    .addTargets([addTargetSchema('sum by (source_zone) (rate(%s[30s]))' % rgwMetric, 1, 'time_series', '{{source_zone}}')]) + {gridPos: {x: x, y: y, w: w, h: h}};
+
+    dashboardSchema('RGW Sync Overview', 'rgw-sync-overview', 'now-1h', '15s', 16, ["overview"], '', {refresh_intervals:['5s','10s','15s','30s','1m','5m','15m','30m','1h','2h','1d'],time_options:['5m','15m','1h','6h','12h','24h','2d','7d','30d']})
+
+    .addAnnotation(addAnnotationSchema(1, '-- Grafana --', true, true, 'rgba(0, 211, 255, 1)', 'Annotations & Alerts', 'dashboard'))
+
+    .addRequired(type='grafana', id='grafana', name='Grafana', version='5.0.0')
+
+    .addRequired(type='panel', id='graph', name='Graph', version='5.0.0')
+
+    .addTemplate(addTemplateSchema('rgw_servers', '$datasource', 'prometehus', 1, 2, true, 1))
+
+    .addTemplate(grafana.template.datasource('datasource', 'prometheus', 'default', label='Data Source'))
+
+    .addPanels([
+    RgwSyncOverviewPanel('Replication (throughput) from Source Zone', 'Bps', null, 'ceph_data_sync_from_zone_fetch_bytes_sum', 0, 0, 8, 7),
+    RgwSyncOverviewPanel('Replication (objects) from Source Zone', 'short', 'Objects/s', 'ceph_data_sync_from_zone_fetch_bytes_count', 8, 0, 8, 7),
+    RgwSyncOverviewPanel('Polling Request Latency from Source Zone', 'ms', null, 'ceph_data_sync_from_zone_poll_latency_sum', 16, 0, 8, 7),
+    RgwSyncOverviewPanel('Unsuccessful Object Replications from Source Zone', 'short', 'Count/s', 'ceph_data_sync_from_zone_fetch_errors', 0, 7, 8, 7)
+    ])
```
I hate to be that picky... 🙈 but part of the goal of replacing JSON dashboards with jsonnet was to make them extremely easy to read, review, etc. So, why not applying a reviewer/human-friendly formatting? (This is the output from jsonnetfmt + some manual whitespacing):
```jsonnet
# If we really want to simplify this, we can import grafonnet as 'g'
local g = import '../../../../grafonnet-lib/grafonnet/grafana.libsonnet';

# We don't really need this, as they're just invoked once
/*
local dashboard = grafana.dashboard;
local template = grafana.template;
local graphPanel = grafana.graphPanel;
local prometheus = grafana.prometheus;
local annotation = grafana.annotation;
local tablePanel = grafana.tablePanel;
*/

local dashboardSchema(title, uid, time_from, refresh, schemaVersion, tags,timezone, timepicker) =
  g.dashboard.new(title=title, uid=uid, time_from=time_from, refresh=refresh, schemaVersion=schemaVersion, tags=tags, timezone=timezone, timepicker=timepicker);

local graphPanelSchema(title, nullPointMode, stack, formatY1, formatY2, labelY1, labelY2, min, fill, datasource) =
  g.graphPanel.new(title=title, nullPointMode=nullPointMode, stack=stack, formatY1=formatY1, formatY2=formatY2, labelY1=labelY1, labelY2=labelY2, min=min, fill=fill, datasource=datasource);

local addTargetSchema(expr, intervalFactor, format, legendFormat) =
  g.prometheus.target(expr=expr, intervalFactor=intervalFactor, format=format, legendFormat=legendFormat);

local addTemplateSchema(name, datasource, query, refresh, hide, includeAll, sort) =
  g.template.new(name=name, datasource=datasource, query=query, refresh=refresh, hide=hide, includeAll=includeAll, sort=sort);

local addAnnotationSchema(builtIn, datasource, enable, hide, iconColor, name, type) =
  g.grafana.annotation.datasource(builtIn=builtIn, datasource=datasource, enable=enable, hide=hide, iconColor=iconColor, name=name, type=type);

{
  "radosgw-sync-overview.json":
    local RgwSyncOverviewPanel(title, formatY1, labelY1, rgwMetric, x, y, w, h) =
      graphPanelSchema(title, 'null as zero', true, formatY1, 'short', labelY1, null, 0, 1, '$datasource')
      .addTargets(
        [addTargetSchema('sum by (source_zone) (rate(%s[30s]))' % rgwMetric, 1, 'time_series', '{{source_zone}}')]) + {gridPos: {x: x, y: y, w: w, h: h}};

    dashboardSchema(
      'RGW Sync Overview', 'rgw-sync-overview', 'now-1h', '15s', 16, ["overview"], '', {refresh_intervals:['5s','10s','15s','30s','1m','5m','15m','30m','1h','2h','1d'],time_options:['5m','15m','1h','6h','12h','24h','2d','7d','30d']}
    )
    .addAnnotation(
      addAnnotationSchema(
        1, '-- Grafana --', true, true, 'rgba(0, 211, 255, 1)', 'Annotations & Alerts', 'dashboard')
    )
    .addRequired(
       type='grafana', id='grafana', name='Grafana', version='5.0.0'
    )
    .addRequired(
       type='panel', id='graph', name='Graph', version='5.0.0'
    )
    .addTemplate(
       addTemplateSchema('rgw_servers', '$datasource', 'prometehus', 1, 2, true, 1)
    )
    .addTemplate(
       g.template.datasource('datasource', 'prometheus', 'default', label='Data Source')
    )
    .addPanels([
      RgwSyncOverviewPanel(
        'Replication (throughput) from Source Zone', 'Bps', null, 'ceph_data_sync_from_zone_fetch_bytes_sum', 0, 0, 8, 7),
      RgwSyncOverviewPanel(
        'Replication (objects) from Source Zone', 'short', 'Objects/s', 'ceph_data_sync_from_zone_fetch_bytes_count', 8, 0, 8, 7),
      RgwSyncOverviewPanel(
        'Polling Request Latency from Source Zone', 'ms', null, 'ceph_data_sync_from_zone_poll_latency_sum', 16, 0, 8, 7),
      RgwSyncOverviewPanel(
        'Unsuccessful Object Replications from Source Zone', 'short', 'Count/s', 'ceph_data_sync_from_zone_fetch_errors', 0, 7, 8, 7)
    ])
```

I'm not

tchaikov commented on src/pybind/mgr/dashboard/CMakeLists.txt:
```diff
@@ -4,12 +4,20 @@ install(DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}
 
+include(ExternalProject)
+ExternalProject_Add(grafonnet-lib
+  DOWNLOAD_DIR ${PROJECT_SOURCE_DIR}
```
please do not create files in `${PROJECT_SOURCE_DIR}`. also this option only applies to "URL Download". see https://cmake.org/cmake/help/latest/module/ExternalProject.html

but you are using Git here.

tchaikov commented on src/pybind/mgr/dashboard/CMakeLists.txt:
```diff
@@ -4,12 +4,20 @@ install(DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}
   if(WITH_TESTS)
     include(AddCephTest)
-    add_tox_test(mgr-dashboard TOX_ENVS py3 lint check openapi-check)
+    include_directories(grafonnet-lib ${PROJECT_SOURCE_DIR}})
```
please refer to https://cmake.org/cmake/help/latest/command/include_directories.html

this line does not make sense at all.

i'd add `grafonnet-lib` as a dependency of `tests`, so it can be prepared before running `ctest`. like
```cmake
add_dependencies(tests
  grafonnet-lib)
```

as an alternative, we can create a fixture for `run-tox-mgr-dashboard`, like
```cmake
if(WITH_TESTS)
  include(AddCephTest)
  add_tox_test(mgr-dashboard
    TOX_ENVS py3 lint check openapi-check grafonnet-check)
  set(ver 0.1.0)
  set(name grafonnet-lib)
  set(tar ${CMAKE_CURRENT_BINARY_DIR}/v${ver}.tar.gz)
  file(DOWNLOAD 
    https://github.com/grafana/grafonnet-lib/archive/refs/tags/v{ver}.tar.gz
    ${tar})
  add_test(
    NAME extract-${name}
    COMMAND ${CMAKE_COMMAND} -E tar xfz ${tar}
    WORKING_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR})
  set_tests_properties(extract-${${name} PROPERTIES
    FIXTURES_SETUP ${name})
  add_test(
    NAME remove-${name}
    COMMAND ${CMAKE_COMMAND} -E remove_directory ${name}-${ver})
  set_tests_properties(remove-${name} PROPERTIES
    FIXTURES_CLEANUP ${name})
  set_property(
    TEST run-tox-mgr-dashboard
    APPEND_STRING
    PROPERTY FIXTURES_REQUIRED ${name})
```

aaSharma14 replied:
@tchaikov , It still gives the error while importing grafonnet-lib - 'couldn't open import "../../../../grafonnet-lib/grafonnet/grafana.libsonnet. Tried both the methods you suggested.

tchaikov replied:
#42623

aaSharma14 commented:
jenkins test make check

aaSharma14 commented:
jenkins test make check

tchaikov commented on monitoring/grafana/dashboards/test-jsonnet.sh:
```diff
@@ -0,0 +1,30 @@
+
+if [[ $(wc -l < ${TEMPDIR}/json_difference.log) -eq 0 ]]
+then
+        rm -rf ${TEMPDIR}
```
please be consistent and use 4 spaces for indent.

tchaikov commented on ceph.spec.in:
```diff
@@ -330,6 +330,8 @@ BuildRequires:  rdma-core-devel
 # for prometheus-alerts
 BuildRequires:  golang-github-prometheus-prometheus
+# for grafana dashboards
+BuildRequires:  jsonnet
```
do we use `jsonnet` if we don't run `make check`, in other words, if user only wants to build ceph is `jsonnet` required? if not, i'd guard by `%if 0%{with make_check}`.

aaSharma14 replied:
Done

tchaikov commented on ceph.spec.in:
```diff
@@ -353,6 +355,7 @@ BuildRequires:	lz4-devel >= 1.7
 %if 0%{with make_check}
 %if 0%{?fedora} || 0%{?rhel}
 BuildRequires:	golang-github-prometheus
+BuildRequires:	golang-github-google-jsonnet
```
it's confusing which implementation of `jsonnet` are we using, the C++ version or the golang version, or both? if both, why?

aaSharma14 replied:
We can keep the golang one

ktdreyer replied:
The C++ one is in EPEL 8, but the golang one is not.

Is `./install-deps.sh` working for you on CentOS 8 now? It's failing for me,
```
No matching package to install: 'golang-github-google-jsonnet'
```

alfonsomthd replied:
@ktdreyer right.
@epuertat @aaSharma14 Our ceph-dev nightly build failed:
https://github.com/rhcs-dashboard/ceph-dev/runs/3308793339?check_suite_focus=true

epuertat replied:
Thanks @ktdreyer ! @aaSharma14 has already fixed that too: #42766

tchaikov commented on src/pybind/mgr/dashboard/CMakeLists.txt:
```diff
@@ -3,13 +3,30 @@ install(DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}
     include(AddCephTest)
-    add_tox_test(mgr-dashboard TOX_ENVS py3 lint check openapi-check)
+    add_tox_test(mgr-dashboard
+        TOX_ENVS grafonnet-check)
```
```suggestion
      TOX_ENVS grafonnet-check)
```

aaSharma14 replied:
Done

tchaikov reviewed (changes requested):
is `src/pybind/mgr/dashboard/requirements-grafonnet.txt` used anywhere?

tchaikov commented on src/pybind/mgr/dashboard/tox.ini:
```diff
@@ -176,6 +181,21 @@ commands =
+passenv = GRAFONNET_PATH
+commands =
+     check: bash {env:CEPH_ROOT:../../../..}/monitoring/grafana/dashboards/test-jsonnet.sh
+     fix: jsonnet -m . {env:CEPH_ROOT:../../../..}/monitoring/grafana/dashboards/jsonnet/grafana_dashboards.jsonnet
```
this tox should be located under `monitoring/grafana/dashboards`.

epuertat replied:
Yup, why not splitting this and moving it to the monitoring/... dir?

aaSharma14 replied:
@tchaikov . I created a new tox file in the grafana/dashboards directory and moved this tox there. But I couldn't see it running in the make check. Do I need to configure it somewhere else as well?

epuertat replied:
@aaSharma14 if you check the top dir's CMakeLists.txt, you'll see that the `monitoring...` dir is conditionally added if `WITH_GRAFANA` option is enabled (and by default is not):

https://github.com/ceph/ceph/blob/a62770a75292267f9ba33430a64dcc831db1a336/CMakeLists.txt#L685-L688

This is because that was for delivering and installing Grafana Dashboards along with the packages, before cephadm and delivering a Grafana containers with the Dashboard already pushed in.

I'd remove from this CMakeLists.txt the previous code that installed the JSON Dashboards, and simply enable the WITH_GRAFANA option in our run-make-check.sh script.

aaSharma14 commented:
> is `src/pybind/mgr/dashboard/requirements-grafonnet.txt` used anywhere?

Yes, I have added the `jsondiff` dependency in this file for the grafonnet-check

tchaikov commented on debian/control:
```diff
@@ -24,6 +24,7 @@ Build-Depends: automake,
                g++ (>= 7),
                javahelper,
                jq <pkg.ceph.check>,
+               jsonnet,
```
```suggestion
               jsonnet <pkg.ceph.check>,
```

aaSharma14 replied:
Done

tchaikov commented on src/pybind/mgr/dashboard/CMakeLists.txt:
```diff
@@ -17,4 +17,4 @@ else()
   install(FILES frontend/package.json
     DESTINATION ${CEPH_INSTALL_DATADIR}/mgr/dashboard/frontend)
-endif()
+endif()
```
i'd keeping the ending newline.

aaSharma14 replied:
Done

tchaikov commented on monitoring/grafana/dashboards/test-jsonnet.sh:
```diff
@@ -0,0 +1,28 @@
+        rm -rf ${TEMPDIR}
+        echo "Grafonnet Check Failed"
+        exit 1
+fi
```
i concur with @p-se . apart from silencing "No newline at end of file" warning, adding a newline at the end of file has a subtile benefit -- we can add a new code lines at the end of this source file with one less line change.

epuertat reviewed (approved):
LGTM! Congrats @aaSharma14. You finally made it!

BTW, it takes ~18 secs (5 secs + 13 secs for setup):

![image](https://user-images.githubusercontent.com/37327689/129061895-3996e503-2079-40ec-a071-f767229a5830.png)

BTW, the Dashboard e2e failure is due to a mismatch with dynamic content in the landing page. @aaryanporwal could you please [have a look at it](https://eyes.applitools.com/app/test-results/00000251773608860102/00000251773608859808/steps/1?accountId=hCTedwnSTUWSCZlXPqua0g~~&mode=step-editor)?

tchaikov commented:
@epuertat @aaSharma14 this change broke the build

```
Processing files: ceph-grafana-dashboards-17.0.0-6884.g034de1f5.el8.noarch
error: Directory not found: /home/jenkins-build/build/workspace/ceph-dev-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/17.0.0-6884-g034de1f5/rpm/el8/BUILDROOT/ceph-17.0.0-6884.g034de1f5.el8.x86_64/etc/grafana/dashboards/ceph-dashboard
error: File not found: /home/jenkins-build/build/workspace/ceph-dev-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/17.0.0-6884-g034de1f5/rpm/el8/BUILDROOT/ceph-17.0.0-6884.g034de1f5.el8.x86_64/etc/grafana/dashboards/ceph-dashboard/*
Executing(%doc): /bin/sh -e /var/tmp/rpm-tmp.EAiLqt
+ umask 022
+ cd /home/jenkins-build/build/workspace/ceph-dev-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/17.0.0-6884-g034de1f5/rpm/el8/BUILD
+ cd ceph-17.0.0-6884-g034de1f5
+ DOCDIR=/home/jenkins-build/build/workspace/ceph-dev-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/17.0.0-6884-g034de1f5/rpm/el8/BUILDROOT/ceph-17.0.0-6884.g034de1f5.el8.x86_64/usr/share/doc/ceph-grafana-dashboards
+ export LC_ALL=C
+ LC_ALL=C
+ export DOCDIR
+ /usr/bin/mkdir -p /home/jenkins-build/build/workspace/ceph-dev-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/17.0.0-6884-g034de1f5/rpm/el8/BUILDROOT/ceph-17.0.0-6884.g034de1f5.el8.x86_64/usr/share/doc/ceph-grafana-dashboards
+ cp -pr monitoring/grafana/dashboards/README /home/jenkins-build/build/workspace/ceph-dev-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/17.0.0-6884-g034de1f5/rpm/el8/BUILDROOT/ceph-17.0.0-6884.g034de1f5.el8.x86_64/usr/share/doc/ceph-grafana-dashboards
+ cp -pr monitoring/grafana/README.md /home/jenkins-build/build/workspace/ceph-dev-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/17.0.0-6884-g034de1f5/rpm/el8/BUILDROOT/ceph-17.0.0-6884.g034de1f5.el8.x86_64/usr/share/doc/ceph-grafana-dashboards
+ exit 0


RPM build errors:
    Directory not found: /home/jenkins-build/build/workspace/ceph-dev-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/17.0.0-6884-g034de1f5/rpm/el8/BUILDROOT/ceph-17.0.0-6884.g034de1f5.el8.x86_64/etc/grafana/dashboards/ceph-dashboard
    File not found: /home/jenkins-build/build/workspace/ceph-dev-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/17.0.0-6884-g034de1f5/rpm/el8/BUILDROOT/ceph-17.0.0-6884.g034de1f5.el8.x86_64/etc/grafana/dashboards/ceph-dashboard/*
+ rm -fr /tmp/install-deps.891037
Build step 'Execute shell' marked build as failure
```

see https://jenkins.ceph.com/job/ceph-dev-build/ARCH=x86_64,AVAILABLE_ARCH=x86_64,AVAILABLE_DIST=centos8,DIST=centos8,MACHINE_SIZE=gigantic/48127//consoleFull

ideepika commented:
https://github.com/ceph/ceph/pull/42765

epuertat commented:
> @epuertat @aaSharma14 this change broke the build
> 
> ```
> Processing files: ceph-grafana-dashboards-17.0.0-6884.g034de1f5.el8.noarch
> error: Directory not found: /home/jenkins-build/build/workspace/ceph-dev-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/17.0.0-6884-g034de1f5/rpm/el8/BUILDROOT/ceph-17.0.0-6884.g034de1f5.el8.x86_64/etc/grafana/dashboards/ceph-dashboard
> error: File not found: /home/jenkins-build/build/workspace/ceph-dev-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/17.0.0-6884-g034de1f5/rpm/el8/BUILDROOT/ceph-17.0.0-6884.g034de1f5.el8.x86_64/etc/grafana/dashboards/ceph-dashboard/*
> Executing(%doc): /bin/sh -e /var/tmp/rpm-tmp.EAiLqt
> + umask 022
> + cd /home/jenkins-build/build/workspace/ceph-dev-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/17.0.0-6884-g034de1f5/rpm/el8/BUILD
> + cd ceph-17.0.0-6884-g034de1f5
> + DOCDIR=/home/jenkins-build/build/workspace/ceph-dev-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/17.0.0-6884-g034de1f5/rpm/el8/BUILDROOT/ceph-17.0.0-6884.g034de1f5.el8.x86_64/usr/share/doc/ceph-grafana-dashboards
> + export LC_ALL=C
> + LC_ALL=C
> + export DOCDIR
> + /usr/bin/mkdir -p /home/jenkins-build/build/workspace/ceph-dev-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/17.0.0-6884-g034de1f5/rpm/el8/BUILDROOT/ceph-17.0.0-6884.g034de1f5.el8.x86_64/usr/share/doc/ceph-grafana-dashboards
> + cp -pr monitoring/grafana/dashboards/README /home/jenkins-build/build/workspace/ceph-dev-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/17.0.0-6884-g034de1f5/rpm/el8/BUILDROOT/ceph-17.0.0-6884.g034de1f5.el8.x86_64/usr/share/doc/ceph-grafana-dashboards
> + cp -pr monitoring/grafana/README.md /home/jenkins-build/build/workspace/ceph-dev-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/17.0.0-6884-g034de1f5/rpm/el8/BUILDROOT/ceph-17.0.0-6884.g034de1f5.el8.x86_64/usr/share/doc/ceph-grafana-dashboards
> + exit 0
> 
> 
> RPM build errors:
>     Directory not found: /home/jenkins-build/build/workspace/ceph-dev-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/17.0.0-6884-g034de1f5/rpm/el8/BUILDROOT/ceph-17.0.0-6884.g034de1f5.el8.x86_64/etc/grafana/dashboards/ceph-dashboard
>     File not found: /home/jenkins-build/build/workspace/ceph-dev-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/17.0.0-6884-g034de1f5/rpm/el8/BUILDROOT/ceph-17.0.0-6884.g034de1f5.el8.x86_64/etc/grafana/dashboards/ceph-dashboard/*
> + rm -fr /tmp/install-deps.891037
> Build step 'Execute shell' marked build as failure
> ```
> 
> see [jenkins.ceph.com/job/ceph-dev-build/ARCH=x86_64,AVAILABLE_ARCH=x86_64,AVAILABLE_DIST=centos8,DIST=centos8,MACHINE_SIZE=gigantic/48127//consoleFull](https://jenkins.ceph.com/job/ceph-dev-build/ARCH=x86_64,AVAILABLE_ARCH=x86_64,AVAILABLE_DIST=centos8,DIST=centos8,MACHINE_SIZE=gigantic/48127//consoleFull)

@tchaikov @ideepika as discussed in the chat, @aaSharma14 is reverting a small subset of changes (installing [Grafana json files in the CMakeLists.txt](https://github.com/ceph/ceph/pull/42194/files#diff-90fe51edeba2655428fd5bbeb383b9b2c9b9a393bda2b9355f205eac9046609aL1-L8)) to see if that's the reason for the failure.

smithfarm commented:
@epuertat @aaSharma14 This PR causes the build to reach out onto the Internet and attempt to download a git repo from github. In environments like the openSUSE Build Service where the build machine does not have a network connection, that completely breaks the build.

Since we already use the `git submodule` feature for including code from other git repos, I wanted to ask whether this external git project could be brought in that way, instead of trying (and failing) to download it at build time?

smithfarm commented:
Also, whatever the fix is, it will need to be backported to pacific, because this PR has already landed in pacific via https://github.com/ceph/ceph/pull/42812 :-/

p-se commented:
Since the compiled Grafana dashboards are included in the repository anyway and the part of the build process that requires a working internet connection seems to be a test only (provided that is correct), would it be viable to disable `WITH_GRAFANA` by default? Changes to those files would still be tested by make check when added to the repository but wouldn't require a working internet connection to be build.

smithfarm commented:
> would it be viable to disable WITH_GRAFANA by default

Hm, wouldn't that break the build of the ceph-grafana-dashboards subpackage, though? See https://github.com/ceph/ceph/commit/5debf36b378b

smithfarm commented:
In general, I believe any cmake code that introduces `ExternalProject_Add` should be guarded appropriately to avoid breaking the standard DEB, RPM builds on machines that aren't connected to the Internet.

There are two ways to download code from the Internet:

1. git submodule
2. make-dist

The latter (`make-dist`) is a script in the top-level directory of the Ceph git repo. The purpose of this script is to generate the tarball from which the DEB,RPM packages are built.

p-se commented:
> > would it be viable to disable WITH_GRAFANA by default
> 
> Hm, wouldn't that break the build of the ceph-grafana-dashboards subpackage, though? See [5debf36](https://github.com/ceph/ceph/commit/5debf36b378b)

Indeed, it does look so. Next idea would be to change `WITH_GRAFANA` (only the occurrences of it in this PR) to `WITH_GRAFANA_TESTS` or and only use that for the make check but, by default, leave it out for building.

smithfarm commented:
> Next idea would be to change WITH_GRAFANA (only the occurrences of it in this PR) to
> WITH_GRAFANA_TESTS or similar and only use that for the make check but, by default,
> leave it out for building.

Yes. Or, just in general, guard any cmake `ExternalProject_Add` directives appropriately to avoid breaking the standard DEB, RPM builds on machines that aren't connected to the Internet.
